File: src/db/entries.ts
import type { Insertable } from 'kysely'
import type { Entries } from 'kysely-codegen'
import type { Database } from './client'

export async function create(db: Database, data: Insertable<Entries>) {
	return db
		.insertInto('entries')
		.values(data)
		.onConflict((oc) =>
			oc.column('link').doUpdateSet({
				title: (eb) => eb.ref('excluded.title'),
				published_at: (eb) => eb.ref('excluded.published_at'),
				description: (eb) => eb.ref('excluded.description'),
				author: (eb) => eb.ref('excluded.author'),
				thumbnail_url: (eb) => eb.ref('excluded.thumbnail_url'),
			}),
		)
		.returning('id')
		.executeTakeFirstOrThrow()
}

File: rs/crates/postgres/src/queries/entries.rs
use colette_database::entries::InsertData;
use sqlx::PgExecutor;

pub async fn insert(ex: impl PgExecutor<'_>, data: InsertData<'_>) -> Result<i64, sqlx::Error> {
    let row = sqlx::query!(
        "
   INSERT INTO entries (link, title, published_at, description, author, thumbnail_url)
   VALUES ($1, $2, $3, $4, $5, $6)
       ON CONFLICT (link) DO
   UPDATE
      SET title = excluded.title,
          published_at = excluded.published_at,
          description = excluded.description,
          author = excluded.author,
          thumbnail_url = excluded.thumbnail_url
RETURNING id",
        data.link,
        data.title,
        data.published_at,
        data.description,
        data.author,
        data.thumbnail_url
    )
    .fetch_one(ex)
    .await?;

    Ok(row.id)
}
